elasticluster/providers/libcloud_provider.py

```
@@ -0,0 +1,210 @@
+#! /usr/bin/env python
+#
+#   Copyright (C) 2013-2017 S3IT, University of Zurich
+#
+#   This program is free software: you can redistribute it and/or modify
+#   it under the terms of the GNU General Public License as published by
+#   the Free Software Foundation, either version 3 of the License, or
+#   (at your option) any later version.
+#
+#   This program is distributed in the hope that it will be useful,
+#   but WITHOUT ANY WARRANTY; without even the implied warranty of
+#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#   GNU General Public License for more details.
+#
+#   You should have received a copy of the GNU General Public License
+#   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+import os
+from tempfile import NamedTemporaryFile
+
+import paramiko
+from libcloud.compute.base import NodeAuthSSHKey, NodeAuthPassword
+from libcloud.compute.providers import get_driver
+from libcloud.compute.types import NodeState
+from paramiko import SSHException
+
+from elasticluster import AbstractCloudProvider
+from elasticluster import log
+from elasticluster.exceptions import KeypairError
+
+EXPLICIT_CONFIG = [
+    # 'floating_ip'
+]
+
+
+class LibCloudProvider(AbstractCloudProvider):
+    driver = None
+    project_name = None
+    floating_ip = False
+    rules = None
+    auth = None
+
+    def pop_args(self, **kwargs):
+        if kwargs.get('username'):
+            return [kwargs.pop('username'), kwargs.pop('password')]
+        elif kwargs.get('access_token'):
+            return kwargs.pop('access token')
+        elif kwargs.get('access_id'):
+            return kwargs.pop('access_id'), kwargs.pop('secret_key')
+        elif kwargs.get('service_account_email'):
+            return [kwargs.pop('service_account_email'), kwargs.pop('pem_file')]
```

Same as for `password` above, if `pem_file` is required then the code should raise `ConfigurationError` if it's missing.